tlatools/org.lamport.tlatools/test/tla2sany/parser/OperatorPrecedenceTests.java

```diff
@@ -1,463 +1,460 @@
+   * @return The higher-prec operator child of the given expression.
+   */
+  private static SyntaxTreeNode getHigherPrecOp(SyntaxTreeNode lowerPrecOp) {
+    SyntaxTreeNode[] lowerPrecOpHeirs = lowerPrecOp.getHeirs();
+    switch (lowerPrecOp.getKind()) {
+      case SyntaxTreeConstants.N_PrefixExpr: return lowerPrecOpHeirs[1];
+      case SyntaxTreeConstants.N_InfixExpr:
+        return lowerPrecOpHeirs[0].isKind(SyntaxTreeConstants.N_GeneralId)
+            ? lowerPrecOpHeirs[2]
+            : lowerPrecOpHeirs[0];
+      case SyntaxTreeConstants.N_PostfixExpr: return lowerPrecOpHeirs[0];
+    }
+    Assert.fail();
+    return null;
+  }
+
+  /**
+   * Given an operator expression, retrieves the string representation of
+   * the operator being used.
+   *
+   * @param op The operator expression.
+   * @return String representation of the operator being used.
+   */
+  private static String getOpImage(SyntaxTreeNode op) {
+    SyntaxTreeNode[] opHeirs = op.getHeirs();
+    switch (op.getKind()) {
+      case SyntaxTreeConstants.N_PrefixExpr: return opHeirs[0].getHeirs()[1].getImage();
+      case SyntaxTreeConstants.N_InfixExpr: return opHeirs[1].getHeirs()[1].getImage();
+      case SyntaxTreeConstants.N_PostfixExpr: return opHeirs[1].getHeirs()[1].getImage();
+    }
+    Assert.fail();
+    return null;
+  }
+
+  /**
+   * Checks the parse tree to ensure the operators were parsed as expected.
+   *
+   * @param root The root of the parse tree.
+   * @param op1 The operator appearing first.
+   * @param op1Symbol The symbol of the operator appearing first.
+   * @param op2 The operator appearing second.
+   * @param op2Symbol The symbol of the operator appearing second.
+   */
+  private static void checkParsePrecedence(
+      SyntaxTreeNode root,
+      Operator op1,
+      String op1Symbol,
+      Operator op2,
+      String op2Symbol
+  ) {
+    SyntaxTreeNode lowerPrecOp = getExpressionInModule(root);
+    String lowerPrecOpSymbol = getOpImage(lowerPrecOp);
+    SyntaxTreeNode higherPrecOp = getHigherPrecOp(lowerPrecOp);
+    String higherPrecOpSymbol = getOpImage(higherPrecOp);
+    if (op1 == op2) {
+      if (op1.Associative) {
+        Assert.assertEquals(lowerPrecOpSymbol, op2Symbol);
+        Assert.assertEquals(higherPrecOpSymbol, op1Symbol);
+      } else {
+        Assert.assertEquals(lowerPrecOpSymbol, op1Symbol);
+        Assert.assertEquals(higherPrecOpSymbol, op2Symbol);
+      }
+    } else if (op1.lowerPrecThan(op2) || FixKind.PREFIX == op2.Fix) {
+      Assert.assertEquals(lowerPrecOpSymbol, op1Symbol);
+      Assert.assertEquals(higherPrecOpSymbol, op2Symbol);
+    } else {
+      Assert.assertEquals(lowerPrecOpSymbol, op2Symbol);
+      Assert.assertEquals(higherPrecOpSymbol, op1Symbol);
+    }
+  }
+
+  /**
+   * Generates all test cases.
+   */
+  @Parameters(name = "{index}: {1}, {3}")
+  public static List<Object[]> getTests() {
+    final List<Object[]> testCases = new ArrayList<Object[]>();
```

Comment:
To avoid excessive allocations at startup, consider initializing the ArrayList with a large number of elements, such as 10k (or the number of tests), instead of growing it dynamically.